### PR TITLE
(re)(re)Introducing the libglvnd OpenGL architecture (attempt #3)

### PR DESCRIPTION
**Pull request by stephencrowell**
- Third attempt of introducing `libglvnd` as described [here](https://github.com/spack/spack/pull/17697)
- Introduced EGL interface using `libglvnd`
- Generalized mesa-demos to depend on `gl` instead of `mesa`
- Changed `provides(gl@:x.x, when=@x.x:)` to `provides(gl@:x.x, when=@x.x)` for all versions except most recent inside `opengl`. The former version would not be solvable because of conflicting variants.

### CODE REVIEW

File: lib/spack/docs/getting_started.rst [hunk 1]
```diff
@@ -997,6 +997,90 @@ to add the following to ``packages.yaml``:
...
+          prefix: /
+        - spec: opengl+glx+glvnd@4.5:
+          prefix: /does/not/exist
+        buildable: false
+        variants: +glx+glvnd
+      all:
+        providers:
+          libglvnd-be-gl: [opengl+glvnd]
+          libglvnd-be-glx: [opengl+glx+glvnd]
+
```
Comment (chuckatkins) [inline lib/spack/docs/getting_started.rst]: Add EGL to the documentation example configuration

File: var/spack/repos/builtin/packages/libglvnd-fe/package.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    @property
+    def gl_libs(self):
+        result = find_libraries('libOpenGL',
+                                root=self.spec['libglvnd'].prefix,
+                                shared=True,
+                                recursive=True)
+        print('\nexporting GL LIBS:\n')
+        print('\n')
+        print(result)
+        print('\n')
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/libglvnd-fe/package.py]: Get rid of debug print statements

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 3]
```diff
@@ -64,10 +68,15 @@ class Mesa(MesonPackage):
...
 
     # Provides
-    provides('gl@4.5',  when='+opengl')
-    provides('glx@1.4', when='+glx')
-    # provides('egl@1.5', when='+egl')
-    provides('osmesa', when='+osmesa')
+    provides('gl@4.5',  when='+opengl ~glvnd')
+    provides('glx@1.4', when='+glx ~glvnd')
+    # This isn't possible b/c of conflict. Should remove?
+    # provides('egl@1.5', when='+egl ~glvnd')
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa/package.py]: `egl` can still be provided when `+egl ~glvnd`.  The conflict for `+egl +opengl ~glvnd` will take care of the invalid configuration.

File: etc/spack/defaults/packages.yaml [hunk 4]
```diff
@@ -25,9 +25,13 @@ packages:
       fftw-api: [fftw, amdfftw]
       flame: [libflame, amdlibflame]
       fuse: [libfuse]
-      gl: [mesa+opengl, mesa18, opengl]
       glu: [mesa-glu, openglu]
-      glx: [mesa+glx, mesa18+glx, opengl]
```
Comment (chuckatkins) [inline etc/spack/defaults/packages.yaml:30]: Let's leave `mesa18` in here for now.  It can be deprecated and removed in a separate pr.

File: etc/spack/defaults/packages.yaml [hunk 5]
```diff
@@ -25,9 +25,13 @@ packages:
...
+      gl: [libglvnd-fe, mesa+opengl~glvnd, opengl~glvnd]
+      glx: [libglvnd-fe+glx, mesa+opengl+glx~glvnd, opengl+glx~glvnd]
```
Comment (chuckatkins) [inline etc/spack/defaults/packages.yaml]: `glx` can be provided via `mesa+glx~glvnd`, no need to add `+opengl`

File: etc/spack/defaults/packages.yaml [hunk 6]
```diff
@@ -25,9 +25,13 @@ packages:
...
+      egl: [libglvnd-fe+egl, opengl+egl~glvnd]
```
Comment (chuckatkins) [inline etc/spack/defaults/packages.yaml]: Add `mesa+egl~glvnd`

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 7]
```diff
@@ -204,44 +242,81 @@ def meson_args(self):
...
-        return LibraryList()
+            if libs_to_seek:
+                return find_libraries(list(libs_to_seek),
+                                      root=self.spec.prefix,
+                                      recursive=True)
+        return LibraryList(())
 
     @property
     def osmesa_libs(self):
+        if '+glvnd' in self.spec:
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa/package.py]: `osmesa` can be provided regardless of `glvnd` status

File: etc/spack/defaults/packages.yaml (hunk 6 above)
Comment (chuckatkins) [reply etc/spack/defaults/packages.yaml]:
```
egl: [libglvnd-fe+egl, opengl+egl~glvnd]
```
should be
```
egl: [libglvnd-fe+egl, mesa+egl~opengl, opengl+egl~glvnd]
```

File: etc/spack/defaults/packages.yaml [hunk 8]
```diff
@@ -25,9 +25,13 @@ packages:
       fftw-api: [fftw, amdfftw]
       flame: [libflame, amdlibflame]
       fuse: [libfuse]
-      gl: [mesa+opengl, mesa18, opengl]
       glu: [mesa-glu, openglu]
-      glx: [mesa+glx, mesa18+glx, opengl]
+      gl: [libglvnd-fe, mesa~glvnd, mesa18, opengl~glvnd]
+      glx: [libglvnd-fe+glx, mesa+glx~glvnd, mesa18+glx, opengl+glx~glvnd]
+      egl: [libglvnd-fe+egl, opengl+egl~glvnd]
+      libglvnd-be-gl: [mesa+glvnd, opengl+glvnd]
```
Comment (chuckatkins) [inline etc/spack/defaults/packages.yaml]: If I'm understanding this correctly, shouldn't this be `libglvnd-be-gl: [mesa+opengl+glvnd, opengl+glvnd]`?

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 9]
```diff
@@ -236,12 +283,36 @@ def osmesa_libs(self):
 
     @property
     def glx_libs(self):
+        if '+glvnd' in self.spec:
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa/package.py]: Can we drop the early exit
```python
if '+glvnd' in self.spec:
    return LibraryList()
```
from the `glx_libs`, `egl_libs`, and `gl_libs` properties since the property functions should never end up getting invoked anyways when `+glvnd` is specified?

File: etc/spack/defaults/packages.yaml (hunk 8 above)
Comment (stephencrowell) [reply etc/spack/defaults/packages.yaml]: I thought we wanted to remove `+opengl` from the mesa versions of `libglvnd-be` pacakges. Or should `+opengl` only be removed from mesa versions in `libglvnd-be-glx` and `libglvnd-be-egl`?

Comment (chuckatkins) [reply etc/spack/defaults/packages.yaml]: As discussed in our call, it looks like we can drop the variants from all the provider entries and just list the packages, letting the concretizer enable the necessary variants as appropriate

File: lib/spack/docs/getting_started.rst [hunk 10]
```diff
@@ -997,6 +997,83 @@ to add the following to ``packages.yaml``:
...
+          prefix: /does/not/exist
+          extra_attributes: 
+            glvnd:
+              glx: 'glx_driver_name'
+              egl: '/path/to/egl/vendor'
+        buildable: false
+        variants: +egl+glx+glvnd
+      all:
+        providers:
+          libglvnd-be-gl: [opengl+glvnd]
```
Comment (chuckatkins) [inline lib/spack/docs/getting_started.rst]: See above comment regarding dropping variants from the provider lists

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 11]
```diff
@@ -64,10 +68,14 @@ class Mesa(MesonPackage):
...
-    # provides('egl@1.5', when='+egl')
-    provides('osmesa', when='+osmesa')
+    provides('gl@4.5',  when='+opengl ~glvnd')
+    provides('glx@1.4', when='+glx ~glvnd')
+    provides('egl@1.5', when='+egl ~glvnd')
+
+    provides('libglvnd-be-gl', when='+opengl +glvnd')
+    provides('libglvnd-be-glx', when='+opengl +glvnd +glx')
+    provides('libglvnd-be-egl', when='+opengl +glvnd +egl')
+    provides('osmesa', when='+osmesa +glvnd')
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa/package.py]: ```provides('osmesa', when='+osmesa')```

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 12]
```diff
@@ -78,13 +86,34 @@ class Mesa(MesonPackage):
...
+             'libxshmfence',
+             'libxxf86vm',
+             'xf86vidmodeproto'), ('+egl', '+glvnd')):
+        depends_on(dependency, when=constraint)
+
     # version specific issue
     # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96130
     conflicts('%gcc@10.1.0', msg='GCC 10.1.0 has a bug')
 
+    # Requre glx or egl when glvnd is enabled
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa/package.py]: `s/Requre/Require/`

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 13]
```diff
@@ -204,29 +241,39 @@ def meson_args(self):
...
+
     @property
     def libs(self):
         spec = self.spec
         libs_to_seek = set()
 
-        if '+osmesa' in spec:
-            libs_to_seek.add('libOSMesa')
+        if '~glvnd' in spec:
+            if '+osmesa' in spec:
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa/package.py]: Move the `if '+osmesa' in spec:` to outside the `if '~glvnd':` block

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 14]
```diff
@@ -204,29 +241,39 @@ def meson_args(self):
...
-        if libs_to_seek:
-            return find_libraries(list(libs_to_seek),
-                                  root=self.spec.prefix,
-                                  recursive=True)
-        return LibraryList()
+            if '+opengles' in spec:
+                libs_to_seek.add('libGLES')
+                libs_to_seek.add('libGLES2')
+
+            if libs_to_seek:
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/mesa/package.py]: `if libs_to_seek:` should be out of any of the other `if` blocks

File: var/spack/repos/builtin/packages/opengl/package.py [hunk 15]
```diff
@@ -91,5 +91,34 @@ def fetcher(self):
 
     @property
     def libs(self):
```
Comment (chuckatkins) [inline var/spack/repos/builtin/packages/opengl/package.py:94]: Restructure in a similar way to the mes `def libs(self):` function, including with the properties

**Review by chuckatkins — DISMISSED** [review]

File: var/spack/repos/builtin/packages/opengl/package.py [hunk 16]
```diff
@@ -14,27 +14,27 @@ class Opengl(Package):
...
-    provides('gl@:1.4', when='@1.4:')
-    provides('gl@:1.3', when='@1.3:')
-    provides('gl@:1.2', when='@1.2:')
-    provides('gl@:1.1', when='@1.1:')
-    provides('gl@:1.0', when='@1.0:')
-
-    if sys.platform != 'darwin':
-        provides('glx@1.4')
+    # Note that this is a dummy SHA since the package is strictly external
+    version('4.5', sha256='0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/opengl/package.py]: You can add a version without a checksum.

File: lib/spack/docs/getting_started.rst [hunk 17]
```diff
@@ -997,6 +997,83 @@ to add the following to ``packages.yaml``:
...
+          libglvnd-be-glx: [opengl]
+          libglvnd-be-egl: [opengl]
+
+One final detail about the above example is that it avoids setting the true
+root of the external OpenGL implementation, instead opting to set it to a path
+that is not expected to exist on the system.  This is done for two reasons.
+First, Spack would add directories under this root to environment variables
+that would affect the process of building and installing other packages, such
+as ``PATH`` and ``PKG_CONFIG_PATH``.  These additions may potentially prevent
+those packages from installing successfully, and this risk is especially great
```
Comment (scheibelp) [inline lib/spack/docs/getting_started.rst]: (question) Spack should omit all system paths etc. from `PATH` - were you seeing an issue where some were in fact being added?

File: lib/spack/docs/getting_started.rst [hunk 18]
```diff
@@ -997,6 +997,83 @@ to add the following to ``packages.yaml``:
...
+provide libglvnd versions of these dependencies ("libglvnd-be-gl",
+"libglvnd-be-glx", etc.).  The libglvnd-fe package depends on these
+"libglvnd-be-..." virtual packages, which provide the actual implementation.
+It also depends on libglvnd, itself, and exposes its libraries to downstream
+applications.  For correct operation, the Spack package for the underlying
+implementation has to set the runtime environment to ensure that it is loaded
+when an application linked against libglvnd runs.  This last detail is
+important for users who want to set up an external OpenGL implementation that
+requires libglvnd to work.  This setup requires modifying the ``packages``
+configuration so that the necessary environment variables are set.
```
Comment (scheibelp) [inline lib/spack/docs/getting_started.rst]: The previous two paragraphs are very useful for me reviewing the PR, but IMO this should be simplified for this particular section. I think a more-direct explanation here would be useful:

```
For EGL support, determine the GLX driver name with <command example> and
EGL vendor path with <command-example> and add the following entry to the
configuration...
```

I think this information would be excellent in the PR description, and/or in the docstrings of the associated packages.

File: lib/spack/docs/getting_started.rst [hunk 19]
```diff
@@ -997,6 +997,83 @@ to add the following to ``packages.yaml``:
...
+One final detail about the above example is that it avoids setting the true
+root of the external OpenGL implementation, instead opting to set it to a path
+that is not expected to exist on the system.  This is done for two reasons.
+First, Spack would add directories under this root to environment variables
+that would affect the process of building and installing other packages, such
+as ``PATH`` and ``PKG_CONFIG_PATH``.  These additions may potentially prevent
+those packages from installing successfully, and this risk is especially great
+for paths that house many libraries and applications, like ``/usr``.  Second,
+providing the true root of the external implementation in the ``packages``
+configuration is not necessary because libglvnd need only the environment
```
Comment (scheibelp) [inline lib/spack/docs/getting_started.rst]: (very minor) "need only" -> "needs only" or "only needs"

File: lib/spack/docs/getting_started.rst [hunk 20]
```diff
@@ -997,6 +997,83 @@ to add the following to ``packages.yaml``:
...
+variables set above in the ``packages`` configuration to determine what OpenGL
```
Comment (scheibelp) [inline lib/spack/docs/getting_started.rst]: It wasn't clear to me that the configuration sets environment variables, reading through the packages themselves I see they set environment variables based on glvnd[glx/egl] - IMO it would be helpful to rephrase this.

File: etc/spack/defaults/packages.yaml [hunk 21]
```diff
@@ -25,9 +25,13 @@ packages:
       fftw-api: [fftw, amdfftw]
       flame: [libflame, amdlibflame]
       fuse: [libfuse]
-      gl: [mesa+opengl, mesa18, opengl]
       glu: [mesa-glu, openglu]
-      glx: [mesa+glx, mesa18+glx, opengl]
+      gl: [libglvnd-fe, mesa, mesa18, opengl]
```
Comment (scheibelp) [inline etc/spack/defaults/packages.yaml:29]: You added `provides('gl', when='~glvnd')` to the `opengl` package: I think it would make sense to say:

```
gl: [libglvnd-fe, mesa, mesa18, opengl~glvnd]
```

or did you try that and encounter a problem with it?

**Review by scheibelp — COMMENTED** [review]
I started looking through this and have a few preliminary questions/comments. I'll be able to take a more-thorough look tomorrow.

File: etc/spack/defaults/packages.yaml (hunk 21 above)
Comment (stephencrowell) [reply etc/spack/defaults/packages.yaml:29]: With the new concretizer,  we don't need to specify the variants in the default packages. If we wanted to be clearer,`gl` should say: 

`gl: [libglvnd-fe, mesa~glvnd, mesa18, opengl~glvnd]`

and the `glx` and `egl` packages should have similar modifications. We decided to let the concretizer be as flexible as possible and removed the variants from default packages.

File: lib/spack/docs/getting_started.rst (hunk 17 above)
Comment (stephencrowell) [reply lib/spack/docs/getting_started.rst]: I believe this is an artifact from previous attempts of introducing `libglvnd` to `spack`. After doing some preliminary testing, I am not seeing any issues with changing the `prefix` to `/` or some other directory that exists.

File: etc/spack/defaults/packages.yaml (hunk 21 above)
Comment (scheibelp) [reply etc/spack/defaults/packages.yaml:29]: That makes sense: it's true that the concretizer can handle this since you've encoded that information in the `package.py`. IMO it would be easier to read if also updated here (although it's definitely not necessary).

File: lib/spack/docs/getting_started.rst (hunk 20 above)
Comment (stephencrowell) [reply lib/spack/docs/getting_started.rst]: I added an extra section below the YAML block that explains the two variables set the environment variables and that they should be changed before using.

File: var/spack/repos/builtin/packages/opengl/package.py [hunk 22]
```diff
@@ -91,5 +90,48 @@ def fetcher(self):
...
+
+        if '~glvnd' in spec:
+            if '+glx' in spec:
+                libs_to_seek.add('libGL')
+
+            if '+egl' in spec:
+                libs_to_seek.add('libEGL')
+
+            if '+opengl' in spec:
+                libs_to_seek.add('libGL')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/opengl/package.py:109]: Existing installations using `opengl` will not find any libraries since the `glvnd` variant will not be present. We could add a warning to users that they should reinstall such packages, but I might want to consult with the Spack team internally about how we'd want to approach this issue in general.

File: var/spack/repos/builtin/packages/opengl/package.py [hunk 23]
```diff
@@ -14,27 +14,26 @@ class Opengl(Package):
...
-        provides('glx@1.4')
+    version('4.5')
+
+    is_linux = sys.platform.startswith('linux')
+    variant('glx', default=is_linux, description="Enable GLX API.")
+    variant('egl', default=False, description="Enable EGL API.")
+    variant('glvnd', default=is_linux,
+            description="Expose Graphics APIs through libglvnd")
+
+    provides('gl', when='~glvnd')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/opengl/package.py:27]: We might want to preserve the above constraints like:

```
provides('gl@:2.1', when='@2.1:')
```

Since that helps dependents determine whether they have a sufficiently new implementation of `opengl` (in particlular when requesting it via asking for a particular `gl` version).

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 24]
```diff
@@ -204,6 +241,12 @@ def meson_args(self):
 
         return args
 
+    # NOTE: Each of the following *libs properties return empty lists if
+    # +glvnd, because there are no mesa libraries to be linked against.  When
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/mesa/package.py:245]: I might rephrase this to say `.libs returns an empty list if +glvnd, because...` but I might have misunderstood: for one thing `+osmesa+glvnd` (which appears to be allowed) looks like it would return `libOSMesa` at least

**Review by scheibelp — COMMENTED** [review]
I've looked through this more and have a couple more requests. Also based on reading through these packages I think the following are all valid dependency trees for `mesa-demos`:

```
# Provided virtuals are given in parentheses before the dependency
mesa-demos+osmesa
- (gl, osmesa) mesa+osmesa+opengl~glvnd

mesa-demos+osmesa
- (osmesa) mesa+osmesa+opengl+glvnd
    - libglvnd
- (gl) opengl~glvnd

mesa-demos+osmesa
- (osmesa) mesa+osmesa+opengl+glvnd
    - libglvnd
- (gl) libglvnd-fe
    - libglvnd
    - (libglvnd-be-*) mesa+osmesa+opengl+glvnd
        - libglvnd

mesa-demos~osmesa
- (gl) libglvnd-fe
    - libglvnd
    - (libglvnd-be-*) mesa+opengl+glvnd
        - libglvnd

mesa-demos~osmesa
- (gl) libglvnd-fe
    - libglvnd
    - (libglvnd-be-*) opengl+glvnd

mesa-demos~osmesa
- (gl) mesa+opengl~glvnd
```

That's not comprehensive but I'm trying to get a sense of the overall organization of things. Please let me know if any of those look wrong.

After coming up with these examples I think it would be clearer if `mesa`'s `+opengl` variant were renamed to `+gl`.

File: var/spack/repos/builtin/packages/opengl/package.py (hunk 22 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/opengl/package.py:109]: To follow up here: this should be fine, although a debugging statement about how previously-found libraries may no longer be found would be helpful (i.e. add a `tty.debug` message telling the user to reinstall if none of the new variants are set but a library is present).

File: var/spack/repos/builtin/packages/opengl/package.py (hunk 23 above)
Comment (stephencrowell) [reply var/spack/repos/builtin/packages/opengl/package.py:27]: For 'gl', the version is runtime dependent, not compile time dependent. There is no meaning attached to a version at this level.

**Comment by scheibelp** [comment]
Hi @stephencrowell I wanted to check in: was this ready for another review?

**Comment by stephencrowell** [comment]
@scheibelp Sorry, I was out of office for the holidays. Yes, the PR is ready for review again. Thanks in advance.

**Comment by chuckatkins** [comment]
> Hi @stephencrowell I wanted to check in: was this ready for another review?

@scheibelp Please do yes.  We're trying to get all the GL rework landed this week.  Regarding the GL version information, it's really a runtime API version so I'm not sure it makes sense to specify in the provides.  For example, mesa can provide 4.6 but some drivers will only provide 3.2, etc.  If we really need the version info tough then we can put it in, it will just be somewhat arbitrary.

**Comment by scheibelp** [comment]
Regarding

> Please let me know if any of those look wrong.

in https://github.com/spack/spack/pull/26301#pullrequestreview-823187507 - can you respond to that?

**Comment by scheibelp** [comment]
>> Since that helps dependents determine whether they have a sufficiently new implementation of opengl (in particlular when requesting it via asking for a particular gl version).
  
> For 'gl', the version is runtime dependent, not compile time dependent. There is no meaning attached to a version at this level.

> Regarding the GL version information, it's really a runtime API version so I'm not sure it makes sense to specify in the provides. For example, mesa can provide 4.6 but some drivers will only provide 3.2, etc.

If the drivers are automatically selected at runtime, then it could make sense to specify the highest version that is provided. If this highest version can change between systems/installations, then that is likely something that needs to be accounted for in Spack's model.

**Comment by stephencrowell** [comment]
> I've looked through this more and have a couple more requests. Also based on reading through these packages I think the following are all valid dependency trees for `mesa-demos`:
> 
> ```
> # Provided virtuals are given in parentheses before the dependency
> mesa-demos+osmesa
> - (gl, osmesa) mesa+osmesa+opengl~glvnd
> 
> mesa-demos+osmesa
> - (osmesa) mesa+osmesa+opengl+glvnd
>     - libglvnd
> - (gl) opengl~glvnd
> 
> mesa-demos+osmesa
> - (osmesa) mesa+osmesa+opengl+glvnd
>     - libglvnd
> - (gl) libglvnd-fe
>     - libglvnd
>     - (libglvnd-be-*) mesa+osmesa+opengl+glvnd
>         - libglvnd
> 
> mesa-demos~osmesa
> - (gl) libglvnd-fe
>     - libglvnd
>     - (libglvnd-be-*) mesa+opengl+glvnd
>         - libglvnd
> 
> mesa-demos~osmesa
> - (gl) libglvnd-fe
>     - libglvnd
>     - (libglvnd-be-*) opengl+glvnd
> 
> mesa-demos~osmesa
> - (gl) mesa+opengl~glvnd
> ```
> 
> That's not comprehensive but I'm trying to get a sense of the overall organization of things. Please let me know if any of those look wrong.
> 
> After coming up with these examples I think it would be clearer if `mesa`'s `+opengl` variant were renamed to `+gl`.

We updated mesa-demos to no longer depend on `gl` without `+glx` or `+egl`. Regardless, I believe that all of these dependency trees are valid.

File: lib/spack/docs/getting_started.rst (hunk 17 above)
Comment (scheibelp) [reply lib/spack/docs/getting_started.rst]: Should this ~section~ paragraph be eliminated then, if the issues it describes no longer exists?

Comment (stephencrowell) [reply lib/spack/docs/getting_started.rst]: Should be up to date now.

**Comment by scheibelp** [comment]
Would it also be possible to model this by

- Removing the libglvnd-fe package
- Having opengl+glvnd depend on libglvnd when +glvnd
- Implement gl_libs, glx_libs, etc. in mesa/opengl

 ? I think that might be easier to understand (if it works)

**Comment by stephencrowell** [comment]
> Would it also be possible to model this by
> 
>     * Removing the libglvnd-fe package
> 
>     * Having opengl+glvnd depend on libglvnd when +glvnd
> 
>     * Implement gl_libs, glx_libs, etc. in mesa/opengl
> 
> 
>  ? I think that might be easier to understand (if it works)

I don't see why it wouldn't be possible; however, I am not sure how this makes it easier to understand. While this reduces the number of packages that can provide `gl`,  this pushes all of the complexity inside the `libglvnd-fe` package into `opengl` and `mesa`.

**Comment by scheibelp** [comment]
>  I am not sure how this makes it easier to understand. While this reduces the number of packages that can provide gl, this pushes all of the complexity inside the libglvnd-fe package into opengl and mesa.

My intent is to simplify some of the trees like:

```
mesa-demos+osmesa
- (osmesa) mesa+osmesa+opengl+glvnd
    - libglvnd
- (gl) libglvnd-fe
    - libglvnd
    - (libglvnd-be-*) mesa+osmesa+opengl+glvnd
        - libglvnd
```

into

```
mesa-demos+osmesa
- (osmesa, gl) mesa+osmesa+opengl+glvnd
    - libglvnd
```

Overall, I think the variants which switch `mesa` between a direct and indirect provider are somewhat complex (it took me multiple hours of reading this PR to get a sense of how these were supposed to relate - so minimally I think some explanation would be extremely useful).

In this case, my understanding is that the complexity which would need to be transferred (EDIT) from `libglvnd-fe` to the `mesa`/`opengl` packages is the definition of `*_libs` (once `libglvnd-fe` is removed for example, this would also eliminate the `libglvnd-be-*` virtuals). To me that tradeoff is worthwhile. We might be able to discuss how to enable code reuse between the packages (admittedly it's generally difficult to do this outside of defining a new `Package` subclass in `build_systems`) if that's the primary concern.

**Comment by chuckatkins** [comment]
> Would it also be possible to model this by
> 
>     * Removing the libglvnd-fe package
> 
>     * Having opengl+glvnd depend on libglvnd when +glvnd
> 
>     * Implement gl_libs, glx_libs, etc. in mesa/opengl
> 
> 
>  ? I think that might be easier to understand (if it works)

It is complicated but the other ways I've tried to represent it don't actually work because of dependency inversions.  The basic problem that brings the complexity is that different parts of `libglvnd` sit at different places in the dependency hierarchy.  It provides the libraries and headers needed by an application using OpenGL but it also provides the libraries and headers needed by a runtime GL implementation.  Nothing inherently depends on the runtime implementation though.  Just having the `opengl` and `mesa` packages depend on `libglvnd` breaks the external use case since a stand alone installation of a glvnd-enabled mesa or `opengl` driver only installs the runtime backends for glvnd with none of the headers and stub libraries needed an application.  So we need to have a way to enable an externally provided opengl with a spack-provided libglvnd.  The way we currently have it is the only way I've been able to see it work so that an application get's the headers and stub libraries provided libglvnd and still requires a backend implementation to be available for runtime, which will in turn also depend on `libglvnd` to build.

**Comment by scheibelp** [comment]
> Just having the opengl and mesa packages depend on libglvnd breaks the external use case since a stand alone installation of a glvnd-enabled mesa or opengl driver only installs the runtime backends for glvnd with none of the headers and stub libraries needed an application.

One possibility is to create an analog of `CudaPackage` which adds a dependency on `libglvnd` like:

```
class GlPackage(Package):
  depends_on(gl)
  depends_on(libglvnd, when=^mesa+glvnd)
  depends_on(libglvnd, when=^opengl+glvnd)

  def setup_build_environment(self, env):
    ...
```

and then packages which would declare a depndency on `gl` would need to inherit this package. On one hand, this pushes some complexity to dependents of `gl`. On the other hand, I think this makes it easier to maintain `opengl` and `mesa` (in my opinion a lot easier). And I could consider how to make Spack support this use case more easily (for one, we are working on mechanisms for integrating externals with dependencies, which may be useful for things like this in the future).

**Comment by scheibelp** [comment]
BTW this is complex enough where if you want to discuss it in a meeting I could set one up: I'm hoping my suggestions are straightforward to implement but I may be overlooking something.

**Comment by chuckatkins** [comment]
> > Just having the opengl and mesa packages depend on libglvnd breaks the external use case since a stand alone installation of a glvnd-enabled mesa or opengl driver only installs the runtime backends for glvnd with none of the headers and stub libraries needed an application.
> 
> One possibility is to create an analog of `CudaPackage` which adds a dependency on `libglvnd`

That would push a fair ammount more complexity onto the consuming application's package, which I'd really like to avoid.  The implementation here keeps the complexity restricted to the underlying gl stack in just a hand full of packages instead of requiring every consumer of opengl to be changed.  We could simplify it some to use glvnd *always* when on linux instead of enabling it as a variant.  This should be doable, even for `mesa18`, since vendor binary drivers have been providing both direct and glvnd-indirect  libraries with their installations for years now.

It looks like we also had a large documentation section that got left out.  Would that address your concerns: adding the documentation and removing glvnd as a variant to just always use the indirect implementation?

Note that splitting libglvnd into multiple pseudo and virtual packages is a common approach and how all the primary distro package managers have been dealing with this for a few years now.

**Comment by scheibelp** [comment]
> It looks like we also had a large documentation section that got left out. Would that address your concerns: adding the documentation 

Documentation might help, I'd be interested to read it and get a sense of whether it makes things easier to understand. That being said IMO the overall modeling is complex and I'd like to improve it.

> and removing glvnd as a variant to just always use the indirect implementation?

That would simplify things, I think that would mean

* opengl/mesa are always indirect providers (i.e. `+glvnd` is now what always happens)
* all packages would now require `libglvnd` (EDIT: all packages *depending on gl* would need `libglvnd`)

I'm not sure about the ramifications of the second constraint: would some users potentially not be able to use this on older systems (in other words, what was `~glvnd` for)?

**Comment by alalazo** [comment]
Is this PR still relevant?
